src/osdc/Objecter.cc

```diff
@@ -2969,9 +2973,21 @@ void Objecter::handle_osd_op_reply(MOSDOpReply *m)
     // FIXME: two redirects could race and reorder
 
     op->tid = 0;
-    m->get_redirect().combine_with_locator(op->target.target_oloc,
-					   op->target.target_oid.name);
-    op->target.flags |= CEPH_OSD_FLAG_REDIRECTED;
+    if (m->is_redirect_reply()){
+      m->get_redirect().combine_with_locator(op->target.target_oloc,
+					    op->target.target_oid.name);
+      op->target.flags |= CEPH_OSD_FLAG_REDIRECTED;
+    }
+
+    if (rc == -ERECOVERING){
+      if (op->target.flags & CEPH_OSD_FLAG_BALANCE_READS){
+        op->target.flags &= ~CEPH_OSD_FLAG_BALANCE_READS;
+        op->target.pgid = pg_t();
```

Comment:
in the EAGAIN block below just unconditionally clear the BALANCE_READS flag.  you shouldn't need to touch pgid.